pkg/server/api/entry.go

```diff
@@ -45,3 +47,56 @@ func RegistrationEntryToProto(e *common.RegistrationEntry) (*types.Entry, error)
 		DnsNames:      e.DnsNames,
 	}, nil
 }
+
+// ProtoToRegistrationEntry converts and validate entry into common registration entry
+func ProtoToRegistrationEntry(e *types.Entry) (*common.RegistrationEntry, error) {
+	if e == nil {
+		return nil, errors.New("missing entry")
+	}
+
+	parentID, err := IDFromProto(e.ParentId)
+	if err != nil {
+		return nil, fmt.Errorf("invalid parent ID: %v", err)
+	}
+
+	spiffeID, err := IDFromProto(e.SpiffeId)
+	if err != nil {
+		return nil, fmt.Errorf("invalid spiffe ID: %v", err)
+	}
+
+	for _, dnsName := range e.DnsNames {
+		if err := x509util.ValidateDNS(dnsName); err != nil {
+			return nil, fmt.Errorf("invalid DNS name: %v", err)
+		}
+	}
+
+	// Validate and normalize TDs
+	for i, federatedWith := range e.FederatesWith {
+		td, err := spiffeid.TrustDomainFromString(federatedWith)
+		if err != nil {
+			return nil, fmt.Errorf("invalid federated trust domain: %v", err)
+		}
+		e.FederatesWith[i] = td.IDString()
+	}
+
+	var selectors []*common.Selector
+	for _, s := range e.Selectors {
+		selectors = append(selectors, &common.Selector{
+			Type:  s.Type,
```

Comment:
Do we need to do any validation on these fields since they are generic strings?

Reply:
I was thinking about that but we dont have any kind of validations about selectors, maybe @azdagron  can have some feedback there?

Reply:
I think some validation here is a good idea. We've had hard-to-debug attestation failures because somebody passed a selector with `{type: "k8s:sa" value:"my-security-group"}` instead of `{type:"k8s" value:"sa:my-security-group"}`.

I'd be ok with some loose validation that ensures there is no `:` in the type. I'm not sure about validation beyond that.